File: truss-compiler/src/test/java/truss/compiler/preprocessor/PreProcessorFixture.java
package truss.compiler.preprocessor;

import org.apache.commons.io.IOUtils;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import truss.compiler.TestCase;
import truss.compiler.TestCaseCallback;

import java.io.InputStream;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;

import static org.junit.Assert.assertEquals;

@RunWith(JUnit4.class)
public class PreProcessorFixture implements TestCaseCallback {
    @Test
    public void withoutDirectives() throws Exception {
        test("WithoutDirectives.t3s");
    }

    @Test
    public void expressionDirectives() throws Exception {
        test("ExpressionDirectives.t3s");
    }

    @Test
    public void defineExpressions() throws Exception {
        test("DefineExpressions.t3s");
    }

    @Test
    public void nestedDirectives() throws Exception {
        test("NestedDirectives.t3s");
    }

    @Override
    public void test(String name, String fileName, String code, String expected) throws Exception {
        assertEquals(String.format("Parse of '%s' failed", name), expected, parse(name, code).trim());
    }

    private void test(String resourceName) throws Exception {
        try (InputStream is = getClass().getResourceAsStream(resourceName)) {
            TestCase testCase = null;

            for (String line : IOUtils.readLines(is)) {
                if (line.equals("<<<")) {
                    testCase.beginExpected();
                } else if (line.startsWith(">>>")) {
                    if (testCase != null) {
                        testCase.test(this);
                    }
                    testCase = new TestCase(line.substring(3).trim(), resourceName);
                } else if (testCase != null) {
                    testCase.append(line);
                }
            }

            if (testCase != null) {
                testCase.test(this);
            }
        }
    }

    protected String parse(String name, String code) throws Exception {
        char[] data;
        try {
            data = PreProcessor.process(
                code.toCharArray(),
                code.length(),
                Collections.unmodifiableSet(new HashSet<>(Arrays.asList("DEFINE1", "DEFINE2")))
            );
        } catch (PreProcessException e) {
            return e.getMessage();
        }

        return new String(data);
    }
}

File: Truss.Compiler.Test/PreProcessor/PreProcessorFixture.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using NUnit.Framework;

namespace Truss.Compiler.Test.PreProcessor {
    [TestFixture]
    public class PreProcessorFixture : ITestCaseCallback {
        [Test]
        public void WithoutDirectives() {
            Test("WithoutDirectives.trs");
        }

        [Test]
        public void ExpressionDirectives() {
            Test("ExpressionDirectives.trs");
        }

        [Test]
        public void DefineExpressions() {
            Test("DefineExpressions.trs");
        }

        [Test]
        public void NestedDirectives() {
            Test("NestedDirectives.trs");
        }

        public void Test(String name, String fileName, String code, String expected) {
            Assert.AreEqual(expected, Parse(name, code).Trim(), String.Format("Parse of '{0}' failed", name));
        }

        private void Test(String resourceName) {
            using (var stream = GetType().Assembly.GetManifestResourceStream(GetType().Namespace + "." + resourceName)) {
                using (var reader = new StreamReader(stream)) {
                    TestCase testCase = null;
                    string line;

                    while ((line = reader.ReadLine()) != null) {
                        if (line == "<<<") {
                            testCase.BeginExpected();
                        } else if (line.StartsWith(">>>")) {
                            if (testCase != null) {
                                testCase.Test(this);
                            }
                            testCase = new TestCase(line.Substring(3).Trim(), resourceName);
                        } else if (testCase != null) {
                            testCase.Append(line);
                        }
                    }

                    if (testCase != null) {
                        testCase.Test(this);
                    }
                }
            }
        }

        protected String Parse(String name, String code) {
            char[] data;
            var messages = new MessageCollection();

            using (MessageCollectionScope.Create(messages)) {
                data = Compiler.PreProcessor.PreProcessor.Process(
                    code.ToCharArray(),
                    code.Length,
                    new[] { "DEFINE1", "DEFINE2" }
                );

                if (messages.HasMessages) {
                    return messages.ToString().Trim();
                }
            }

            return new String(data);
        }
    }
}
